packages/pigeon/README.md

```diff
@@ -3,11 +3,341 @@
-It also improves efficiency over common method channel patterns. Most importantly though,
-it removes the need to write custom platform channel code, since pigeon generates it for you.
+Pigeon works by reading a special file or files, which are placed outside of the
+`/lib` directory which hosts all of your application code. You define special
+data classes and endpoints, which Pigeon will then consume and use to generate
+matching Dart and native code at the paths you specify.
 
-For usage examples, see the [Example README](./example/README.md).
+Internally, the generated code uses `MethodChannel`s to communicate between Flutter's
+UI thread and the Platform thread where your host app is initially launched. The value
+in Pigeon comes from automatically keeping this unpleasant boilerplate in sync and
+efficiently marshalling data between languages.
+
+The generated code can either flow from Dart to native code, or native code back
+to Dart. Generated code on the receiving end uses interfaces or abstract classes,
+allowing you to provide implementations in concrete classes.
+
+Pigeon works in both complete Flutter apps and hybrid apps using the add-to-app paradigm.
+
+## Quickstart
+
+### Installation
+
+Begin by adding Pigeon to your Dart project's `pubspec.yaml` file:
+
+```sh
+$ flutter pub add pigeon --dev
+$ flutter pub get
+```
+
+### Setup
+
+To specify what code Pigeon should generate, create your interface definition file.
+This guide will place all such definitions at `/pigeons/messages.dart`, but you
+are free to choose any file name you like, inside a dedicated folder or not.
+
+Begin by instantiating a `PigeonOptions` object, wrapped in the `@ConfigurePigeon`
+decorator. Later, you will pass named parameters to your `PigeonOptions` object
+to specify your desired behavior.
+
+```dart
+// pigeons/messages.dart
+
+import 'package:pigeon/pigeon.dart';
+
+@ConfigurePigeon(
+  PigeonOptions()
+);
+```
+
+### Define your messages
+
+The next step to use Pigeon is to define the data structures Dart will exchange
+with native code. You do this by writing plain Dart enums or classes in your
+definition file. These messages should only contain simple constructors and direct
+attributes. Methods, factory constructors, constructor bodies, and computed
+properties are all not allowed.
+
+```dart
+// pigeons/messages.dart
+
+enum Code { a, b }
+
+class MessageData {
+  MessageData({required this.code, required this.data});
+  String? name;
+  String? description;
+  Code code;
+  Map<String, String> data;
+}
+```
+
+No extra steps are necessary to register these classes - their inclusion
+in your definitions file ensures Pigeon will generate matching Dart and native
+implementations.
+
+### Define which methods to expose
+
+The point of Pigeon and the `MethodChannel`s it utilizes is to call native functions
+living on the Platform thread from Dart, or to call Dart functions living on the UI
```

Comment:
This language seems very confusing to me; functions don't live on threads, calls are on threads. There is no inherent binding between a function and a thread.